### PR TITLE
more cautious gunicorn hooks with tests

### PR DESCRIPTION
**Pull request by snopoke**

### CODE REVIEW

File: testapps/test_gunicorn/tests.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+
+class Server:
+    def __init__(self):
+        self.log = Logger()
+
+
+def setup():
+    # ensure env var not set
+    os.environ.pop('prometheus_multiproc_dir')
```
Comment (dannyroberts) [inline testapps/test_gunicorn/tests.py]: Test failure is:
```
======================================================================
ERROR: test suite for <module 'testapps.test_gunicorn.tests' from '/mnt/commcare-hq-ro/testapps/test_gunicorn/tests.py'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python3.6/site-packages/nose/suite.py", line 210, in run
    self.setUp()
  File "/vendor/lib/python3.6/site-packages/nose/suite.py", line 293, in setUp
    self.setupContext(ancestor)
  File "/vendor/lib/python3.6/site-packages/nose/suite.py", line 316, in setupContext
    try_run(context, names)
  File "/vendor/lib/python3.6/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/mnt/commcare-hq-ro/testapps/test_gunicorn/tests.py", line 23, in setup
    os.environ.pop('prometheus_multiproc_dir')
  File "/usr/local/lib/python3.6/_collections_abc.py", line 795, in pop
    value = self[key]
  File "/usr/local/lib/python3.6/os.py", line 669, in __getitem__
    raise KeyError(key) from None
KeyError: 'prometheus_multiproc_dir'
```

**Review by dannyroberts — APPROVED** [review]
Thanks for adding tests, approved once tests pass

File: corehq/tests/noseplugins/uniformresult.py [hunk 2]
```diff
@@ -32,6 +33,8 @@ def uniform_description(test):
             descriptor.__name__,
             test.arg,
         )
+    if isfunction(test):
+        return test.__name__
```
Comment (snopoke) [inline corehq/tests/noseplugins/uniformresult.py]: @millerdev the build was failing here on `test_on_starting`. Not sure why `test_on_starting` is not of type `FunctionTestCase`. Does this seem like a reasonable solution.
```
======================================================================
ERROR: test suite for <function test_on_starting at 0x7f9820345680>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/skelly/.virtualenvs/hq/lib/python3.7/site-packages/nose/suite.py", line 210, in run
    self.setUp()
  File "/home/skelly/.virtualenvs/hq/lib/python3.7/site-packages/nose/suite.py", line 293, in setUp
    self.setupContext(ancestor)
  File "/home/skelly/.virtualenvs/hq/lib/python3.7/site-packages/nose/suite.py", line 302, in setupContext
    self.config.plugins.startContext(context)
  File "/home/skelly/.virtualenvs/hq/lib/python3.7/site-packages/nose/plugins/manager.py", line 99, in __call__
    return self.call(*arg, **kw)
  File "/home/skelly/.virtualenvs/hq/lib/python3.7/site-packages/nose/plugins/manager.py", line 167, in simple
    result = meth(*arg, **kw)
  File "/home/skelly/src/cchq/corehq/tests/noseplugins/timing.py", line 87, in startContext
    self.end_event("before", context)
  File "/home/skelly/src/cchq/corehq/tests/noseplugins/timing.py", line 64, in end_event
    name = uniform_description(context)
  File "/home/skelly/src/cchq/corehq/tests/noseplugins/uniformresult.py", line 39, in uniform_description
    test._testMethodName
AttributeError: 'function' object has no attribute '_testMethodName'
```

File: testapps/test_gunicorn/tests.py [hunk 3]
```diff
@@ -20,7 +20,10 @@ def __init__(self):
 
 def setup():
     # ensure env var not set
-    os.environ.pop('prometheus_multiproc_dir')
+    try:
+        os.environ.pop('prometheus_multiproc_dir')
+    except KeyError:
+        pass
```
Comment (millerdev) [inline testapps/test_gunicorn/tests.py]:
```suggestion
    os.environ.pop('prometheus_multiproc_dir', None)
```

File: corehq/tests/noseplugins/uniformresult.py (hunk 2 above)
Comment (millerdev) [reply corehq/tests/noseplugins/uniformresult.py]: I agree, nose should have wrapped it in `FunctionTestCase` before it gets to this point in the code.

Your fix seems reasonable, but I would also include the module name
```suggestion
        return "%s:%s" % (test.__module__, test.__name__)
```
Alternately, I think line 27 could be changed to
```py
if isinstance(test, type) or isfunction(test):
```

**Review by millerdev — APPROVED** [review]

**Pull request merged**